_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+* There's also a `sleep()` at the bottom of `calculate_top_10()`.
+  Because polling is used, tasks that run too quickly won't get accurate information.
+* Finally, because polling is per-process, you can't run multiple threads per worker, otherwise you'll get results that combine memory usage of multiple tasks.
+
+## Interpreting the data
+
+What we've learned is that memory usage of `calculate_top_10()` grows with file size; this can let us [characterize the memory requirements for the workload](https://pythonspeed.com/articles/estimating-memory-usage/), creating a model saying "for these file sizes and this level of parallelism we will need this much RAM."
+And that can guide our hardware specifications.
+
+However, we might also notice that the memory usage seems rather high: for a 25MB file, we're using 330MB of RAM to count words.
+Thinking through how an ideal version of this code might work, we ought to be able to process the file with much less memory.
+
+**And that's another way in which `dask-memusage` can be helpful: it can point us at specific code that needs memory usage optimized, at the granularity of a task.**
+If you want to dig down further, I've built [a memory profiler called Fil](https://pythonspeed.com/fil) that is expressly designed for finding peak memory usage.
+Unlike `dask-memusage`, which can be run on production workloads, Fil slows down your execution and has other limitations I'm currently working on (it doesn't support multiple processes, as of March 2021), so for now it's better used for manual profiling.
+
+We can write a little script that only runs on `big.txt`:
+
+```python
+from pathlib import Path
+from example2 import calculate_top_10
+
+calculate_top_10(Path("files/big.txt"))
+```
+
+Run it under Fil:
+
+```shell-session
+$ pip install filprofiler
+$ fil-profile run example3.py
+```
+
+And the result shows us where the bulk of the memory is being allocated:
+
+<iframe id="peak" src="/images/dask_memusage/peak-memory.svg" width="100%" height="300" scrolling="auto" frameborder="0"></iframe>
+
+Reading in the file takes 8% of memory, but `data.split()` is responsible for 84% of memory.
+Perhaps we shouldn't be loading the whole file into memory and splitting the whole file into words, and instead we should be processing the file line by line...
+
+## Next steps
+
+If you're running Dask workloads with the Distributed backend, and you're fine with only having one thread per worker, running with `dask-memusage` will give you real-world per-task memory usage on production workloads.
+You can then use this to:
+
+* Fine tune your hardware configuration, so you're not wasting RAM or CPU cores.
+* Or, when possible, you can fine tune your chunking size; smaller chunks will use less memory.
+* As a starting point for optimizing memory usage—you can then use Fil to figure out which lines of code are responsible.
```

Comment:
If this is a good starting point, perhaps it should be the first bullet point in this list